File: sknetwork/utils/check.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created in April 2019
@author: Nathan de Lara <nathan.delara@polytechnique.org>
"""
import warnings
from typing import Union, Optional

import numpy as np
from scipy import sparse


def has_nonnegative_entries(input_matrix: Union[sparse.csr_matrix, np.ndarray]) -> bool:
    """True if the array has non-negative entries."""
    if type(input_matrix) == sparse.csr_matrix:
        return np.all(input_matrix.data >= 0)
    else:
        return np.all(input_matrix >= 0)


def is_weakly_connected(adjacency: sparse.csr_matrix) -> bool:
    """Check whether a graph is weakly connected.
    Parameters
    ----------
    adjacency:
        Adjacency matrix of the graph.
    """
    n_cc = sparse.csgraph.connected_components(adjacency, (not is_symmetric(adjacency)), 'weak', False)
    return n_cc == 1


def check_connected(adjacency: sparse.csr_matrix):
    """Check is a graph is weakly connected and return an error otherwise."""
    if is_weakly_connected(adjacency):
        return
    else:
        raise ValueError('The graph is expected to be connected.')


def check_nonnegative(input_matrix: Union[sparse.csr_matrix, np.ndarray]):
    """Check whether the array has non-negative entries."""
    if not has_nonnegative_entries(input_matrix):
        raise ValueError('Only nonnegative values are expected.')


def has_positive_entries(input_matrix: np.ndarray) -> bool:
    """True if the array has positive entries."""
    if type(input_matrix) != np.ndarray:
        raise TypeError('Entry must be a dense NumPy array.')
    else:
        return np.all(input_matrix > 0)


def check_positive(input_matrix: Union[sparse.csr_matrix, np.ndarray]):
    """Check whether the array has positive entries."""
    if not has_positive_entries(input_matrix):
        raise ValueError('Only positive values are expected.')


def is_proba_array(input_matrix: np.ndarray) -> bool:
    """True if each line of the array has non-negative entries which sum to 1."""
    if len(input_matrix.shape) == 1:
        return has_nonnegative_entries(input_matrix) and np.isclose(input_matrix.sum(), 1)
    elif len(input_matrix.shape) == 2:
        n_row, n_col = input_matrix.shape
        err = input_matrix.dot(np.ones(n_col)) - np.ones(n_row)
        return has_nonnegative_entries(input_matrix) and np.isclose(np.linalg.norm(err), 0)
    else:
        raise TypeError('Entry must be one or two-dimensional array.')


def is_square(input_matrix: Union[sparse.csr_matrix, np.ndarray]) -> bool:
    """True if the matrix is square."""
    return input_matrix.shape[0] == input_matrix.shape[1]


def check_square(input_matrix: Union[sparse.csr_matrix, np.ndarray]):
    """Check whether a matrix is square and return an error otherwise."""
    if is_square(input_matrix):
        return
    else:
        raise ValueError('The adjacency matrix is expected to be square.')


def is_symmetric(input_matrix: sparse.csr_matrix) -> bool:
    """True if the matrix is symmetric."""
    return sparse.csr_matrix(input_matrix - input_matrix.T).nnz == 0


def check_symmetry(input_matrix: sparse.csr_matrix):
    """Check whether a matrix is symmetric and return an error otherwise."""
    if not is_symmetric(input_matrix):
        raise ValueError('The input matrix is expected to be symmetric.')


def make_weights(distribution: str, adjacency: sparse.csr_matrix) -> np.ndarray:
    """Array of weights from a matrix and a desired distribution.

   Parameters
   ----------
   distribution:
       Distribution for node sampling. Only ``'degree'`` or ``'uniform'`` are accepted.
   adjacency:
       The adjacency matrix of the neighbors.

   Returns
   -------
   node_weights: np.ndarray
       Weights of nodes.
    """
    n = adjacency.shape[0]
    distribution = distribution.lower()
    if distribution == 'degree':
        node_weights_vec = adjacency.dot(np.ones(adjacency.shape[1]))
    elif distribution == 'uniform':
        node_weights_vec = np.ones(n)
    else:
        raise ValueError('Unknown distribution of node weights.')
    return node_weights_vec


def check_format(input_matrix: Union[sparse.csr_matrix, sparse.csc_matrix, sparse.coo_matrix, sparse.lil_matrix,
                                     np.ndarray], allow_empty: bool = False) -> sparse.csr_matrix:
    """Check whether the matrix is a NumPy array or a Scipy sparse matrix and return
    the corresponding Scipy CSR matrix.
    """
    formats = {sparse.csr_matrix, sparse.csc_matrix, sparse.coo_matrix, sparse.lil_matrix, np.ndarray}
    if type(input_matrix) not in formats:
        raise TypeError('The input matrix must be in Scipy sparse format or Numpy ndarray format.')
    input_matrix = sparse.csr_matrix(input_matrix)
    if not allow_empty and input_matrix.nnz == 0:
        raise ValueError('The input matrix is empty.')
    return input_matrix


def check_is_proba(entry: Union[float, int], name: str = None):
    """Check whether the number is non-negative and less than or equal to 1."""
    if name is None:
        name = 'Probabilities'
    if type(entry) not in [float, int]:
        raise TypeError('{} must be floats (or ints if 0 or 1).'.format(name))
    if entry < 0 or entry > 1:
        raise ValueError('{} must be between 0 and 1.'.format(name))


def check_damping_factor(damping_factor: float):
    """Check if the damping factor has a valid value."""
    if damping_factor < 0 or damping_factor >= 1:
        raise ValueError('A damping factor must have a value in [0, 1[.')


def check_weights(weights: Union['str', np.ndarray], adjacency: Union[sparse.csr_matrix, sparse.csc_matrix],
                  positive_entries: bool = False) -> np.ndarray:
    """Check whether the weights are a valid distribution for the adjacency and return a probability vector.

    Parameters
    ----------
    weights:
        Probabilities for node sampling in the null model. ``'degree'``, ``'uniform'`` or custom weights.
    adjacency:
        The adjacency matrix of the graph.
    positive_entries:
        If true, the weights must all be positive, if False, the weights must be non-negative.

    Returns
    -------
    node_weights: np.ndarray
        Valid weights of nodes.
    """
    n = adjacency.shape[0]
    if type(weights) == np.ndarray:
        if len(weights) != n:
            raise ValueError('The number of node weights must match the number of nodes.')
        else:
            node_weights_vec = weights
    elif type(weights) == str:
        node_weights_vec = make_weights(weights, adjacency)
    else:
        raise TypeError(
            'Node weights must be a known distribution ("degree" or "uniform" string) or a custom NumPy array.')

    if positive_entries and not has_positive_entries(node_weights_vec):
        raise ValueError('All weights must be positive.')
    else:
        if np.any(node_weights_vec < 0) or node_weights_vec.sum() <= 0:
            raise ValueError('Node weights must be non-negative with positive sum.')

    return node_weights_vec


def get_probs(weights: Union['str', np.ndarray], adjacency: Union[sparse.csr_matrix, sparse.csc_matrix],
              positive_entries: bool = False) -> np.ndarray:
    """Check whether the weights are a valid distribution for the adjacency
    and return a normalized probability vector.
    """
    weights = check_weights(weights, adjacency, positive_entries)
    return weights / np.sum(weights)


def check_random_state(random_state: Optional[Union[np.random.RandomState, int]]):
    """Check whether the argument is a seed or a NumPy random state. If None, 'numpy.random' is used by default."""
    if random_state is None:
        return np.random.RandomState()
    elif type(random_state) == int:
        return np.random.RandomState(random_state)
    elif type(random_state) == np.random.RandomState:
        return random_state
    else:
        raise TypeError('To specify a random state, pass the seed (as an int) or a NumPy random state object.')


def check_n_neighbors(n_neighbors: int, n_seeds: int):
    """Set the number of neighbors so that it is less than the number of labeled samples."""
    if n_neighbors >= n_seeds:
        warnings.warn(Warning("The number of neighbors must be lower than the number of nodes with known labels. "
                              "Changed accordingly."))
        n_neighbors = n_seeds - 1
    return n_neighbors


def check_labels(labels: np.ndarray):
    """Check labels of the seeds for semi-supervised algorithms."""

    classes: np.ndarray = np.unique(labels[labels >= 0])
    n_classes: int = len(classes)

    if n_classes < 2:
        raise ValueError('There must be at least two distinct labels.')
    else:
        return classes, n_classes


def check_n_jobs(n_jobs: Optional[int] = None):
    """Parse the ``n_jobs`` parameter for multiprocessing."""
    if n_jobs == -1:
        return None
    elif n_jobs is None:
        return 1
    else:
        return n_jobs


def check_adjacency_vector(adjacency_vectors: Union[sparse.csr_matrix, np.ndarray],
                           n: Optional[int] = None) -> sparse.csr_matrix:
    """Check format of new samples for predict methods"""
    adjacency_vectors = check_format(adjacency_vectors)
    if n is not None and adjacency_vectors.shape[1] != n:
        raise ValueError('The adjacency vector must be of length equal to the number nodes in the graph.')
    return adjacency_vectors


def check_n_clusters(n_clusters: int, n_row: int, n_min: int = 0):
    """Check that the number of clusters"""
    if n_clusters > n_row:
        raise ValueError('The number of clusters exceeds the number of rows.')
    if n_clusters < n_min:
        raise ValueError('The number of clusters must be at least {}.'.format(n_min))
    else:
        return


def check_min_size(n_row, n_min):
    """Check that an adjacency has the required number of rows and returns an error otherwise."""
    if n_row < n_min:
        raise ValueError('The graph must contain at least {} nodes.'.format(n_min))
    else:
        return


def check_dendrogram(dendrogram):
    """Check the shape of a dendrogram."""
    if dendrogram.ndim != 2 or dendrogram.shape[1] != 4:
        raise ValueError("Dendrogram has incorrect shape.")
    else:
        return


def check_min_nnz(nnz, n_min):
    """Check that an adjacency has the required number of edges and returns an error otherwise."""
    if nnz < n_min:
        raise ValueError('The graph must contain at least {} edge(s).'.format(n_min))
    else:
        return


def check_n_components(n_components, n_min) -> int:
    """Check the number of components"""
    if n_components > n_min:
        warnings.warn(Warning("The dimension of the embedding cannot exceed {}. Changed accordingly.".format(n_min)))
        return n_min
    else:
        return n_components


def check_scaling(scaling: float, adjacency: sparse.csr_matrix, regularize: bool):
    """Check the scaling factor"""
    if scaling < 0:
        raise ValueError("The 'scaling' parameter must be non-negative.")

    if scaling and (not regularize) and not is_weakly_connected(adjacency):
        raise ValueError("Positive 'scaling' is valid only if the graph is connected or with regularization."
                         "Call 'fit' either with 'scaling' = 0 or positive 'regularization'.")


def has_boolean_entries(input_matrix: np.ndarray) -> bool:
    """True if the array has boolean entries."""
    if type(input_matrix) != np.ndarray:
        raise TypeError('Entry must be a dense NumPy array.')
    else:
        return input_matrix.dtype == 'bool'


def check_boolean(input_matrix: np.ndarray):
    """Check whether the array has positive entries."""
    if not has_boolean_entries(input_matrix):
        raise ValueError('Only boolean values are expected.')


def check_vector_format(vector_1: np.ndarray, vector_2: np.ndarray):
    """Check whether the inputs are vectors of same length."""
    if len(vector_1.shape) > 1 or len(vector_2.shape) > 1:
        raise ValueError('The arrays must be 1-dimensional.')
    if vector_1.shape[0] != vector_2.shape[0]:
        raise ValueError('The arrays do not have the same length.')


def has_self_loops(input_matrix: sparse.csr_matrix) -> bool:
    """True if each node has a self loop."""
    return all(input_matrix.diagonal().astype(bool))


def add_self_loops(adjacency: sparse.csr_matrix) -> sparse.csr_matrix:
    """Add self loops to adjacency matrix.

    Parameters
    ----------
    adjacency : sparse.csr_matrix
        Adjacency matrix of the graph.

    Returns
    -------
    sparse.csr_matrix
        Adjacency matrix of the graph with self loops.
    """
    n_row, n_col = adjacency.shape

    if is_square(adjacency):
        adjacency = sparse.diags(np.ones(n_col), format='csr') + adjacency
    else:
        tmp = sparse.eye(n_row)
        tmp.resize(n_row, n_col)
        adjacency += tmp

    return adjacency

File: sknetwork-rs/src/utils/check.rs
use ndarray::{Array1, Array2};
use sprs::{CsMat, TriMat};

use crate::utils::format::MatrixInput;

#[derive(Debug, Clone, PartialEq, Eq)]
/// Errors raised by check error operations.
pub enum CheckError {
    /// Indicates not connected.
    NotConnected,
    /// Indicates not non negative.
    NotNonNegative,
    /// Indicates not positive.
    NotPositive,
    /// Indicates not proba.
    NotProba,
    /// Indicates not square.
    NotSquare,
    /// Indicates not symmetric.
    NotSymmetric,
    /// Indicates unknown distribution.
    UnknownDistribution,
    /// Indicates wrong weights length.
    WrongWeightsLength,
    /// Indicates invalid weights.
    InvalidWeights,
    /// Indicates invalid proba range.
    InvalidProbaRange,
    /// Indicates invalid damping factor.
    InvalidDampingFactor,
    /// Indicates invalid adjacency vector length.
    InvalidAdjacencyVectorLength,
    /// Indicates invalid nclusters.
    InvalidNClusters,
    /// Indicates invalid min size.
    InvalidMinSize,
    /// Indicates invalid min nnz.
    InvalidMinNnz,
    /// Indicates invalid scaling.
    InvalidScaling,
    /// Indicates invalid vector dimension.
    InvalidVectorDimension,
    /// Indicates invalid vector length.
    InvalidVectorLength,
}

#[derive(Debug, Clone, PartialEq)]
/// WeightsInput enum.
pub enum WeightsInput {
    /// Indicates distribution.
    Distribution(String),
    /// Indicates vector.
    Vector(Array1<f64>),
}

/// Returns whether the input satisfies `has nonnegative entries sparse`.
pub fn has_nonnegative_entries_sparse(input_matrix: &CsMat<f64>) -> bool {
    input_matrix.data().iter().all(|x| *x >= 0.0)
}

/// Returns whether the input satisfies `has nonnegative entries dense`.
pub fn has_nonnegative_entries_dense(input_matrix: &Array2<f64>) -> bool {
    input_matrix.iter().all(|x| *x >= 0.0)
}

/// Returns whether the input satisfies `has positive entries`.
pub fn has_positive_entries(input_matrix: &Array1<f64>) -> bool {
    input_matrix.iter().all(|x| *x > 0.0)
}

/// Validates nonnegative sparse.
///
/// # Errors
///
/// Returns [`CheckError`] on failure.
pub fn check_nonnegative_sparse(input_matrix: &CsMat<f64>) -> Result<(), CheckError> {
    if has_nonnegative_entries_sparse(input_matrix) {
        Ok(())
    } else {
        Err(CheckError::NotNonNegative)
    }
}

/// Validates positive.
///
/// # Errors
///
/// Returns [`CheckError`] on failure.
pub fn check_positive(input_matrix: &Array1<f64>) -> Result<(), CheckError> {
    if has_positive_entries(input_matrix) {
        Ok(())
    } else {
        Err(CheckError::NotPositive)
    }
}

/// Returns whether the input satisfies `is proba array 1d`.
pub fn is_proba_array_1d(input_matrix: &Array1<f64>) -> bool {
    let sum: f64 = input_matrix.sum();
    input_matrix.iter().all(|x| *x >= 0.0) && (sum - 1.0).abs() <= 1e-12
}

/// Returns whether the input satisfies `is proba array 2d`.
pub fn is_proba_array_2d(input_matrix: &Array2<f64>) -> bool {
    for row in input_matrix.rows() {
        if row.iter().any(|x| *x < 0.0) {
            return false;
        }
        let s: f64 = row.iter().sum();
        if (s - 1.0).abs() > 1e-12 {
            return false;
        }
    }
    true
}

/// Returns whether the input satisfies `is square`.
pub fn is_square(input_shape: (usize, usize)) -> bool {
    input_shape.0 == input_shape.1
}

/// Validates square.
///
/// # Errors
///
/// Returns [`CheckError`] on failure.
pub fn check_square(input_shape: (usize, usize)) -> Result<(), CheckError> {
    if is_square(input_shape) {
        Ok(())
    } else {
        Err(CheckError::NotSquare)
    }
}

/// Returns whether the input satisfies `is symmetric`.
pub fn is_symmetric(input_matrix: &CsMat<f64>) -> bool {
    if !is_square(input_matrix.shape()) {
        return false;
    }
    let t = input_matrix.transpose_view().to_owned();
    let diff = input_matrix - &t;
    diff.nnz() == 0
}

/// Validates symmetry.
///
/// # Errors
///
/// Returns [`CheckError`] on failure.
pub fn check_symmetry(input_matrix: &CsMat<f64>) -> Result<(), CheckError> {
    if is_symmetric(input_matrix) {
        Ok(())
    } else {
        Err(CheckError::NotSymmetric)
    }
}

/// Returns whether the input satisfies `is weakly connected`.
pub fn is_weakly_connected(adjacency: &CsMat<f64>) -> bool {
    let (n, m) = adjacency.shape();
    if n == 0 || m == 0 {
        return true;
    }
    let sym = adjacency + &adjacency.transpose_view().to_owned();
    let mut seen = vec![false; n];
    let mut stack = vec![0usize];
    seen[0] = true;
    while let Some(u) = stack.pop() {
        if let Some(row) = sym.outer_view(u) {
            for v in row.indices() {
                if !seen[*v] {
                    seen[*v] = true;
                    stack.push(*v);
                }
            }
        }
    }
    seen.into_iter().all(|x| x)
}

/// Validates connected.
///
/// # Errors
///
/// Returns [`CheckError`] on failure.
pub fn check_connected(adjacency: &CsMat<f64>) -> Result<(), CheckError> {
    if is_weakly_connected(adjacency) {
        Ok(())
    } else {
        Err(CheckError::NotConnected)
    }
}

/// Computes make weights.
///
/// # Errors
///
/// Returns [`CheckError`] on failure.
pub fn make_weights(distribution: &str, adjacency: &CsMat<f64>) -> Result<Array1<f64>, CheckError> {
    let n = adjacency.rows();
    match distribution.to_lowercase().as_str() {
        "degree" => {
            let mut out = Array1::<f64>::zeros(n);
            for i in 0..n {
                let sum = adjacency
                    .outer_view(i)
                    .map(|row| row.data().iter().sum())
                    .unwrap_or(0.0);
                out[i] = sum;
            }
            Ok(out)
        }
        "uniform" => Ok(Array1::<f64>::ones(n)),
        _ => Err(CheckError::UnknownDistribution),
    }
}

/// Validates is proba.
///
/// # Errors
///
/// Returns [`CheckError`] on failure.
pub fn check_is_proba(entry: f64) -> Result<(), CheckError> {
    if (0.0..=1.0).contains(&entry) {
        Ok(())
    } else {
        Err(CheckError::InvalidProbaRange)
    }
}

/// Validates damping factor.
///
/// # Errors
///
/// Returns [`CheckError`] on failure.
pub fn check_damping_factor(damping_factor: f64) -> Result<(), CheckError> {
    if (0.0..1.0).contains(&damping_factor) {
        Ok(())
    } else {
        Err(CheckError::InvalidDampingFactor)
    }
}

/// Validates weights.
pub fn check_weights(
    weights: WeightsInput,
    adjacency: &CsMat<f64>,
    positive_entries: bool,
) -> Result<Array1<f64>, CheckError> {
    let n = adjacency.rows();
    let node_weights = match weights {
        WeightsInput::Vector(w) => {
            if w.len() != n {
                return Err(CheckError::WrongWeightsLength);
            }
            w
        }
        WeightsInput::Distribution(d) => make_weights(&d, adjacency)?,
    };

    if positive_entries {
        if !has_positive_entries(&node_weights) {
            return Err(CheckError::InvalidWeights);
        }
    } else if node_weights.iter().any(|x| *x < 0.0) || node_weights.sum() <= 0.0 {
        return Err(CheckError::InvalidWeights);
    }

    Ok(node_weights)
}

/// Returns probs.
pub fn get_probs(
    weights: WeightsInput,
    adjacency: &CsMat<f64>,
    positive_entries: bool,
) -> Result<Array1<f64>, CheckError> {
    let weights = check_weights(weights, adjacency, positive_entries)?;
    let s = weights.sum();
    Ok(weights / s)
}

/// Validates n neighbors.
pub fn check_n_neighbors(n_neighbors: usize, n_seeds: usize) -> usize {
    if n_seeds == 0 {
        0
    } else {
        n_neighbors.min(n_seeds)
    }
}

/// Validates labels.
///
/// # Errors
///
/// Returns [`usize`] on failure.
pub fn check_labels(labels: &Array1<i32>) -> Result<(Vec<i32>, usize), CheckError> {
    let mut classes: Vec<i32> = labels.iter().copied().filter(|x| *x >= 0).collect();
    classes.sort_unstable();
    classes.dedup();
    let n_classes = classes.len();
    if n_classes < 2 {
        Err(CheckError::InvalidWeights)
    } else {
        Ok((classes, n_classes))
    }
}

/// Validates n jobs.
pub fn check_n_jobs(n_jobs: Option<i32>) -> Option<i32> {
    match n_jobs {
        Some(-1) => None,
        None => Some(1),
        Some(x) => Some(x),
    }
}

/// Validates adjacency vector.
pub fn check_adjacency_vector(
    adjacency_vectors: MatrixInput,
    n: Option<usize>,
) -> Result<CsMat<f64>, CheckError> {
    let adjacency = match adjacency_vectors {
        MatrixInput::Dense(dense) => {
            let (r, c) = dense.dim();
            let mut tri = TriMat::<f64>::new((r, c));
            for i in 0..r {
                for j in 0..c {
                    let v = dense[[i, j]];
                    if v != 0.0 {
                        tri.add_triplet(i, j, v);
                    }
                }
            }
            tri.to_csr()
        }
        MatrixInput::Sparse(s) => s,
    };
    if let Some(expected) = n {
        if adjacency.cols() != expected {
            return Err(CheckError::InvalidAdjacencyVectorLength);
        }
    }
    Ok(adjacency)
}

/// Validates n clusters.
///
/// # Errors
///
/// Returns [`CheckError`] on failure.
pub fn check_n_clusters(n_clusters: usize, n_row: usize, n_min: usize) -> Result<(), CheckError> {
    if n_clusters > n_row || n_clusters < n_min {
        Err(CheckError::InvalidNClusters)
    } else {
        Ok(())
    }
}

/// Validates min size.
///
/// # Errors
///
/// Returns [`CheckError`] on failure.
pub fn check_min_size(n_row: usize, n_min: usize) -> Result<(), CheckError> {
    if n_row < n_min {
        Err(CheckError::InvalidMinSize)
    } else {
        Ok(())
    }
}

/// Validates min nnz.
///
/// # Errors
///
/// Returns [`CheckError`] on failure.
pub fn check_min_nnz(nnz: usize, n_min: usize) -> Result<(), CheckError> {
    if nnz < n_min {
        Err(CheckError::InvalidMinNnz)
    } else {
        Ok(())
    }
}

/// Validates n components.
pub fn check_n_components(n_components: usize, n_min: usize) -> usize {
    if n_components > n_min {
        n_min
    } else {
        n_components
    }
}

/// Validates scaling.
pub fn check_scaling(
    scaling: f64,
    adjacency: &CsMat<f64>,
    regularize: bool,
) -> Result<(), CheckError> {
    if scaling < 0.0 {
        return Err(CheckError::InvalidScaling);
    }
    if scaling > 0.0 && !regularize && !is_weakly_connected(adjacency) {
        return Err(CheckError::InvalidScaling);
    }
    Ok(())
}

/// Validates vector format.
pub fn check_vector_format(
    vector_1: &Array1<f64>,
    vector_2: &Array1<f64>,
) -> Result<(), CheckError> {
    if vector_1.len() != vector_2.len() {
        Err(CheckError::InvalidVectorLength)
    } else {
        Ok(())
    }
}

/// Returns whether the input satisfies `has self loops`.
pub fn has_self_loops(input_matrix: &CsMat<f64>) -> bool {
    let n = input_matrix.rows().min(input_matrix.cols());
    (0..n).any(|i| input_matrix.get(i, i).is_some())
}

/// Computes add self loops.
pub fn add_self_loops(adjacency: &CsMat<f64>) -> CsMat<f64> {
    let (n_row, n_col) = adjacency.shape();
    let mut tri = TriMat::<f64>::new((n_row, n_col));
    for (i, row) in adjacency.outer_iterator().enumerate() {
        for (j, v) in row.iter() {
            tri.add_triplet(i, j, *v);
        }
    }
    for i in 0..n_row.min(n_col) {
        tri.add_triplet(i, i, 1.0);
    }
    tri.to_csr()
}

#[cfg(test)]
mod tests {
    use ndarray::{Array1, array};

    use super::*;

    fn csr_from_dense(dense: &Array2<f64>) -> CsMat<f64> {
        let (r, c) = dense.dim();
        let mut tri = TriMat::<f64>::new((r, c));
        for i in 0..r {
            for j in 0..c {
                let v = dense[[i, j]];
                if v != 0.0 {
                    tri.add_triplet(i, j, v);
                }
            }
        }
        tri.to_csr()
    }

    #[test]
    fn test_square_connected_symmetry() {
        assert!(check_square((3, 3)).is_ok());
        assert!(check_square((3, 4)).is_err());

        let disconnected =
            csr_from_dense(&array![[0.0, 1.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, 0.0]]);
        assert!(check_connected(&disconnected).is_err());

        let symmetric = csr_from_dense(&array![[0.0, 1.0], [1.0, 0.0]]);
        assert!(check_symmetry(&symmetric).is_ok());
    }

    #[test]
    fn test_weights_and_probs() {
        let adjacency = csr_from_dense(&array![[0.0, 2.0], [1.0, 0.0]]);
        let degree = make_weights("degree", &adjacency).unwrap();
        assert_eq!(degree.to_vec(), vec![2.0, 1.0]);
        let probs = get_probs(
            WeightsInput::Distribution("uniform".to_string()),
            &adjacency,
            false,
        )
        .unwrap();
        assert!((probs.sum() - 1.0).abs() <= 1e-12);
    }

    #[test]
    fn test_misc_checks() {
        assert!(check_is_proba(0.5).is_ok());
        assert!(check_damping_factor(0.85).is_ok());
        assert_eq!(check_n_neighbors(10, 5), 5);
        assert_eq!(check_n_neighbors(1, 5), 1);
        assert_eq!(check_n_neighbors(5, 0), 0);

        let labels = Array1::from_vec(vec![0, 1, -1, 1]);
        let (_, n_classes) = check_labels(&labels).unwrap();
        assert_eq!(n_classes, 2);

        let a = Array1::from_vec(vec![1.0, 2.0]);
        let b = Array1::from_vec(vec![3.0, 4.0]);
        assert!(check_vector_format(&a, &b).is_ok());
    }

    #[test]
    fn test_self_loops() {
        let adjacency = csr_from_dense(&array![[0.0, 1.0], [1.0, 1.0]]);
        assert!(has_self_loops(&adjacency));
        let with_loops = add_self_loops(&adjacency);
        assert!(has_self_loops(&with_loops));
    }
}
